agent/agent-gc-monitor/gc-monitor-api/src/main/java/com/microsoft/gcmonitor/memorypools/MemoryPools.java

```diff
@@ -0,0 +1,345 @@
+    PS_EDEN_SPACE(PSEdenSpace.POOL_NAME, PSEdenSpace::new),
+    PS_OLD_GEN(PSOldGen.POOL_NAME, PSOldGen::new),
+    PS_PERM_GEN(PSPermGen.POOL_NAME, PSPermGen::new),
+    PS_SURVIVOR_SPACE(PSSurvivorSpace.POOL_NAME, PSSurvivorSpace::new),
+    CMS_OLD_GEN(CMSOldGen.POOL_NAME, CMSOldGen::new),
+    CMS_PERM_GEN(CMSPermGen.POOL_NAME, CMSPermGen::new),
+    PAR_EDEN_SPACE(ParEdenSpace.POOL_NAME, ParEdenSpace::new),
+    PAR_SURVIVOR_SPACE(ParSurvivorSpace.POOL_NAME, ParSurvivorSpace::new),
+    G1_EDEN(G1Eden.POOL_NAME, G1Eden::new),
+    G1_EDEN_SPACE(G1EdenSpace.POOL_NAME, G1EdenSpace::new),
+    G1_SURVIVOR(G1Survivor.POOL_NAME, G1Survivor::new),
+    G1_SURVIVOR_SPACE(G1SurvivorSpace.POOL_NAME, G1SurvivorSpace::new),
+    G1_OLD_GEN(G1OldGen.POOL_NAME, G1OldGen::new),
+    G1_PERM_GEN(G1PermGen.POOL_NAME, G1PermGen::new),
+    EDEN_SPACE(EdenSpace.POOL_NAME, EdenSpace::new),
+    SURVIVORS_PACE(SurvivorSpace.POOL_NAME, SurvivorSpace::new),
+    TENURED_GEN(TenuredGen.POOL_NAME, TenuredGen::new),
+    PERM_GEN_SHARED_RW(PermGenSharedRw.POOL_NAME, PermGenSharedRw::new),
+    PERM_GEN_SHARED_RO(PermGenSharedRo.POOL_NAME, PermGenSharedRo::new),
+    PERM_GEN(PermGen.POOL_NAME, PermGen::new),
+    METASPACE(Metaspace.POOL_NAME, Metaspace::new),
+    COMPRESSED_CLASS_SPACE(CompressedClassSpace.POOL_NAME, CompressedClassSpace::new),
+    SHENANDOAH(Shenandoah.POOL_NAME, Shenandoah::new),
+    Z_HEAP(ZHeap.POOL_NAME, ZHeap::new);
+
+    interface MemoryPoolFactory extends Function<Set<GarbageCollector>, MemoryPool> {
+    }
+
+    private final String poolName;
+    private final MemoryPoolFactory factory;
+
+    MemoryPools(String poolName, MemoryPoolFactory factory) {
+        this.poolName = poolName;
+        this.factory = factory;
+    }
+
+    public static Optional<MemoryPools> findPoolFor(String poolName) {
+        return Arrays.stream(MemoryPools
+                .values())
+                .filter(pool -> pool.poolName.equals(poolName))
+                .findFirst();
+    }
+
+    public static MemoryPool getMemoryPool(MBeanServerConnection connection, ObjectName name, Set<GarbageCollector> collectors) {
+        MemoryPoolMXBean dataSource = JMX.newMXBeanProxy(connection, name, MemoryPoolMXBean.class);
+        String id = getIdentifier(dataSource.getName());
+
+        Set<GarbageCollector> managers = Arrays
+                .stream(dataSource.getMemoryManagerNames())
+                .flatMap(memoryName -> collectors
+                        .stream()
+                        .filter(col -> col.getName().equals(memoryName)))
+                .collect(Collectors.toSet());
+
+        Optional<MemoryPools> pool = MemoryPools.findPoolFor(id);
+
+        if (pool.isPresent()) {
+            return pool.get().factory.apply(managers);
+        } else {
+            throw new IllegalArgumentException("Cannot find memory pool " + id);
+        }
+    }
+
+    public static String getIdentifier(String name) {
```

Comment:
can you add a comment why we need an id and cannot use name directly?

Reply:
this is a remnant of when we used this as a key inside JSON, it is not needed anymore, can remove it